Fixes #1526 to atomic write metadata operation

suvadityamuk opened the pull request:
Signed-off-by: Suvaditya Mukherjee <suvadityamuk@gmail.com>

Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

Fixes #1526 

**Description of the changes being introduced by the pull request**:

This is an atomic-write operation where a temp file is first written to and atomically re-placed on top of the old original file.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
new API: test containers for zero or more elements

Test metadata (de)serialization with input data containing containers
with zero or more elements.

Here is the status for the different use cases:
Root keys:
- many keys: added
Root roles:
- many roles: added
Root role keyids:
- many keids: already added in https://github.com/theupdateframework/tuf/pull/1481
MetaFile hashes:
- many hashes: already tested
- zero hashes: added. Testing as invalid test case.
Timestamp meta:
- zero elements: already tested
- many elements: added
Snapshot meta:
- zero items: added
- many items: added
Delegation keys:
- many keys: added
Delegation role keyids:
- many keyids: added
Delegation role paths:
- many paths: already tested
Delegation role path_hash_prefixes:
- many path_hash_path_prefixes: already tested
Delegation roles:
- zero roles: added
- multiple roles: added
Targets targets:
- zero items: already tested
- multiple items: added

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Metadata API: include target path in targetfile

Currently, TargetFile instances do not contain the path relative URL of
the file they represent. The API itself does not need it but it could be
useful for users of the API.

As an example, the current client returns a dict for
get_one_valid_targetinfo(): that dict contains a filepath field and
a targetinfo field (essentially TargetFile).
We would like to keep a similar API, but avoid hand-crafted dicts.
It would be much nicer to return a TargetFile that would contain the
full "metadata" of the targetfile.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

sechkova committed:
Add sphinx docs source files

Add .rst source files for building documentation with
'sphinx'. The two mandatory files are conf.py containing
the build configuration and the master doc file index.rst.

Sphinx uses 'autodoc' to automatically include docstrings.
'autodoc' imports the modules and needs TUF installed in
the environment.

The following command will generate the documentation from the
source files in an html format:

`sphinx-build -b html docs/sphinx/source docs/sphinx/build/html`

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Add tox:docs environment

 - New 'docs' environment in tox enables
   building the sphinx documentation in isolation.
 - New requirements-docs.txt.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Add docs/sphinx/build to .gitignore

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Fix sphinx-build warnings

Fix docstrings which triggered warnings from sphinx-build.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Add .readthedocs.yaml

Add a configuration file for Read The Docs.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

Commit:
Remove filename argument from Targets.update()

After the addition of "path" argument in the TargetFile class the
filename argument in Targets.update() became redundant.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
ngclient: Remove root_update_finished()

The usefulness was debatable to begin with, and now that it has become
clear that rollback protection requires a second "final verification"
step for all three root, timestamp and snapshot it is clear that
root_update_finished() is not good design.

update_root() still accepts expired root metadata but now the final
root expiry is checked when the "next" metadata (timestamp) is loaded.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
ngclient: Remove root_update_finished()

The usefulness was debatable to begin with, and now that it has become
clear that rollback protection requires a second "final verification"
step for all three root, timestamp and snapshot it is clear that
root_update_finished() is not good design.

update_root() still accepts expired root metadata but now the final
root expiry is checked when the "next" metadata (timestamp) is loaded.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Additional tests for containers with 0 or >1 items

Those tests are needed to cover use cases when syntatcticly as
standalone objects the metadata classes and their helper classes defined
in tuf/api/metadata.py are valid even if they cannot be verified.

An example where an object is valid, but cannot be verified is
if we have a Role instance with an empty list of "keyids".
This instance is valid and can be created, but cannot be verified
because there is a requirement that the threshold should be above
1, meaning that there should be at least 1 element inside the "keyids"
list to complete successful threshold verification.

The situation is the same for the rest of the tests I am adding to this
commit:
- Root object without keys
- Root object without roles
- DelegationRole object with empty "keyids"
- DelegationRole object with an empty list of "paths"
- DelegationRole object with an empty list of "path_hash_prefixes"
all of these objects can be instantiated, but cannot complete
successfully threshold verification.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
docs: API Reference tweaks

* Improve content
* Make ngclient Updater __init__() visible in docs
* Remove "legacy implementation" (except for the note on API stability)

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
docs: Tweak INSTALLATION.rst

* Remove link to outdated roadmap
* Link to maintainers file in the same way as two lines earlier
* Fix formatting issues with code blocks

These fixes allow the installation rst to be used from sphinx sources
and from docs root.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
docs: Move the sphinx root to docs/

This allows using existing documentation in the published documentation
without
* moving the existing docs (which would break external links)
* tricks like symlinks that create issues with relative links

Put the api reference files into a subdirectory to avoid polluting the
main docs/ directory.

Include "Installation" and "Instructions for Contributors" in the
published documentation.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
docs: Improve tuf.api docs

Write a bit more about the two modules, hide the actual TOC to not
repeat (and not have sphinx complain about missing items in TOC)

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
docs: Improve ngclient docs

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

suvadityamuk committed:
Fixes #1526 to atomic write metadata operation

Signed-off-by: Suvaditya Mukherjee <suvadityamuk@gmail.com>

suvadityamuk committed:
Fixes #1526 to atomic write metadata operation

Signed-off-by: Suvaditya Mukherjee <suvadityamuk@gmail.com>

trishankatdatadog commented on tuf/ngclient/updater.py:
```diff
@@ -292,8 +293,15 @@ def _load_local_metadata(self, rolename: str) -> bytes:
+        if data writing process is halted or interrupted, the original data is not lost.
+        """
+        
+        original_filename = open(f"{rolename}.json", 'a+')
```
I'm not sure we support f-strings yet. Maybe best to stick to our current convention... which is String.format()? @jku

suvadityamuk replied:
The original code that I pulled my first fork out of used `f-strings` in place, that is why I thought this would be the correct way to go about it. 
When I looked at the tests and where they failed, it seems my mistake is the fact that I am passing an `iowrapper` to the `NamedTemporaryFile` function where it expects a `str`, so I think that change could possibly solve the thing. 
But I really do not understand why is it that these errors do not come up when I run `tox` or the `aggregate_tests.py` functions on my system.

trishankatdatadog commented on tuf/ngclient/updater.py:
```diff
@@ -292,8 +293,15 @@ def _load_local_metadata(self, rolename: str) -> bytes:
+        """
+        
+        original_filename = open(f"{rolename}.json", 'a+')
+        file_out = tempfile.NamedTemporaryFile(dir=os.path.dirname(original_filename), delete=False)
```
Hmm, don't we want to clean up after temp files after we are done moving?

suvadityamuk replied:
If there is any problem or the original file is somehow corrupted during any operation, the temp file will act as backup. That's the idea behind keeping the temp file intact, as I can see.

suvadityamuk committed:
Fix to #1536

Signed-off-by: Suvaditya Mukherjee <suvadityamuk@gmail.com>

Checked for linting errors and file handling permissions

Signed-off-by: Suvaditya Mukherjee <suvadityamuk@gmail.com>

Fixes added as expected for #1526

Signed-off-by: Suvaditya Mukherjee <suvadityamuk@gmail.com>

All fixes incorporated as asked to solve #1526

Signed-off-by: Suvaditya Mukherjee <suvadityamuk@gmail.com>

Fixes #1526 final with all suggested changes added

dependabot[bot] committed:
build(deps): bump cryptography from 3.4.7 to 3.4.8

Bumps [cryptography](https://github.com/pyca/cryptography) from 3.4.7 to 3.4.8.
- [Release notes](https://github.com/pyca/cryptography/releases)
- [Changelog](https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst)
- [Commits](https://github.com/pyca/cryptography/compare/3.4.7...3.4.8)

---
updated-dependencies:
- dependency-name: cryptography
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

jku commented:
You can use the same PR for multiple revisions of the PR (force-pushing a new version of the same branch is fine by me): this would be better so we don't lose discussion in the previous PRs

some remaining issues:
 * there's still multiple commits in the branch for a single change
 * I don't understand why original_file_touch is opened?
 * you removed "with NamedTemporaryFile" in this version: that means the file descriptor is open when we call os.replace(): this will lead to failures on Windows

I expect something like this would work
```
    with NamedTemporaryFile(dir=self._dir, delete=False) as tmpfile:
        tmpfile.write(data)
    os.replace(tmpfile.name, file_name)
```

the gains I'm hoping to get are:
 * NamedTemporaryFile selects a safe filename, tmpfile is on same filesystem as file_name
 * "with" closes the tmpfile file descriptor
 * os.replace should work even on windows because there are now no open file descriptors
 * after replace tmpfile does not exist anymore
 * most importantly the actual file can only be either original content or complete new content after os.replace
 
 so the only issue I see is that if tmpfile.write or os.replace fails, the tempfile probably still exists: but i'm fine with that side effect

jku commented:
The _Run TUF tests and linter / build (3.x, ubuntu-latest, with-sslib-master, true)_ failure is not your fault btw, it's an unrelated issue

Commit:
Remove duplicate serialization tests from test_api

I was looking at how can we simplify or split test_api.py when I noticed
that the test cases covered by those two test functions are already
covered in the test_metadata_serialization.py module in the
"invalid_keys" and "invalid_roles" datasets.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

suvadityamuk commented:
That error was scary, thanks for letting me know. The others, I have made the change. I did not know all too well about the Windows side-effects so thank you once again for that. 
I had originally thought that there was some case where the file was not created at all and hence it was unable to detect or overwrite it or something along those lines.  Seems like I was pretty off there as well.

sechkova committed:
docs: Add links to the specification

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
docs: Improve formating

Remove/add new lines at the end of file.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

suvadityamuk commented:
Hey, do check and let me know how I can better this?

Commit:
tox: Fix with-sslib-master install

pip nowadays recognizes that we are asking for two different versions of
securesystemslib in the "with-sslib-master" env, and errors out.

Instead install normal dependencies first, then install the new
securesystemslib separately (this ends up upgrading securesystemslib).

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

sechkova committed:
docs: Fix DelegatedRole rendering

Improve class DelegatedRole docstring in order to be rendered
correctly in the documentation built with sphinx and autodoc.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

jku reviewed (changes requested):
remaining issues:
* there's still multiple commits in the branch for a single change: I'd rather see all squashed into one commit. If you are having trouble with this you may want read up on  "interactive rebase" and "squashing": https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History (and as I mentioned, force pushing a modified branch is fine by me)
* the final filename/path has been changed from `os.path.join(self._dir, f"{rolename}.json")` to `f"{rolename}.json"`: this is incorrect

the tests really _should_ have failed on that second issue, but it looks like we don't test for this currently (the ngclient updater tests are a work in progress)... I did not understand what was happening myself so had to write a quick test to figure it out: it's in https://github.com/jku/tuf/commit/1250890ff7ba25fd0bbb39910c67d9fa0c53385e : feel free to use that change (or if that feels difficult, I can combine your PR and this commit once the PR is otherwise good)

dependabot[bot] committed:
build(deps): bump securesystemslib[crypto,pynacl] from 0.20.1 to 0.21.0

Bumps [securesystemslib[crypto,pynacl]](https://github.com/secure-systems-lab/securesystemslib) from 0.20.1 to 0.21.0.
- [Release notes](https://github.com/secure-systems-lab/securesystemslib/releases)
- [Changelog](https://github.com/secure-systems-lab/securesystemslib/blob/master/CHANGELOG.md)
- [Commits](https://github.com/secure-systems-lab/securesystemslib/compare/v0.20.1...v0.21.0)

---
updated-dependencies:
- dependency-name: securesystemslib[crypto,pynacl]
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

suvadityamuk commented:
@jku Is this fine?

MVrachev commented:
Thanks @suvadityamuk that you continue the pr refinement!
Could you please squash the two commits that are left into one?
When you do, that please rewrite the commit message of the resulted commit , so that it doesn't contain merge of the messages together (as it's the case for commit https://github.com/theupdateframework/tuf/pull/1539/commits/eccdf2b97108e40307f8578fb62016364fd1a37d0).

Additionally, please note that you will have to rebase on top of the `develop` branch and fix the conflicting cases.

suvadityamuk committed:
Merge branch 'develop' into develop

suvadityamuk committed:
Fixes #1526

Signed-off-by: Suvaditya Mukherjee <suvadityamuk@gmail.com>

suvadityamuk commented:
Hi. I think I have rebased the current commits on top of the `develop` branch, but I am really unable to figure out how do I squash the previous commits because they have somehow just cleared off and I cannot find them in the `git rebase -i HEAD`. Is there any chance you could excuse this or something of that sort?
